File: frontend/src/lib/currency.ts
/**
 * Currency formatting (§24).
 *
 * Mirrors backend/app/core/currency.py. Prices are ALWAYS shown in the
 * currency they were recorded in; nothing is converted, because converting a
 * 2021 French sale into pounds at today's rate would produce a number that
 * never existed.
 */

const SYMBOLS: Record<string, string> = {
  GBP: "£", EUR: "€", USD: "$", AUD: "A$", CAD: "C$", NZD: "NZ$",
  INR: "₹", JPY: "¥", CHF: "CHF ", SEK: "kr ", NOK: "kr ", DKK: "kr ",
  ZAR: "R", BRL: "R$", SGD: "S$", HKD: "HK$", PLN: "zł ", CZK: "Kč ",
};

export function symbol(currency: string): string {
  return SYMBOLS[currency?.toUpperCase()] ?? `${currency?.toUpperCase() ?? ""} `;
}

function trim(value: number): string {
  const s = value.toFixed(1);
  return s.endsWith(".0") ? s.slice(0, -2) : s;
}

/** `£425,000` — the exact figure, for headline values. */
export function formatFull(value: number, currency: string): string {
  return `${symbol(currency)}${Math.round(value).toLocaleString("en-GB")}`;
}

/** `£725k`, `£1.2m`, `₹1.8 Cr` — for map markers. */
export function formatCompact(value: number, currency: string): string {
  const cur = currency?.toUpperCase() ?? "";
  const sym = symbol(cur);
  const v = Math.abs(value);

  if (cur === "INR") {
    // Indian numbering: 1 crore = 10,000,000; 1 lakh = 100,000.
    if (v >= 1e7) return `${sym}${trim(v / 1e7)} Cr`;
    if (v >= 1e5) return `${sym}${trim(v / 1e5)} L`;
    return `${sym}${Math.round(v).toLocaleString("en-IN")}`;
  }
  if (cur === "JPY") {
    if (v >= 1e8) return `${sym}${trim(v / 1e8)}億`;
    if (v >= 1e4) return `${sym}${trim(v / 1e4)}万`;
    return `${sym}${Math.round(v).toLocaleString()}`;
  }
  if (v >= 1e9) return `${sym}${trim(v / 1e9)}bn`;
  if (v >= 1e6) return `${sym}${trim(v / 1e6)}m`;
  if (v >= 1e3) return `${sym}${Math.round(v / 1e3)}k`;
  return `${sym}${Math.round(v).toLocaleString()}`;
}

/** `£438,000 – £487,000` */
export function formatRange(
  low: number | null,
  high: number | null,
  currency: string,
): string | null {
  if (low == null || high == null) return null;
  return `${formatFull(low, currency)} – ${formatFull(high, currency)}`;
}

export function formatArea(sqm: number | null): string | null {
  if (sqm == null) return null;
  return `${Math.round(sqm)} m²`;
}

export function formatDate(iso: string | null): string | null {
  if (!iso) return null;
  const d = new Date(iso);
  if (Number.isNaN(d.getTime())) return null;
  return d.toLocaleDateString("en-GB", { day: "numeric", month: "long", year: "numeric" });
}

export function formatMonthYear(iso: string | null): string | null {
  if (!iso) return null;
  const d = new Date(iso);
  if (Number.isNaN(d.getTime())) return null;
  return d.toLocaleDateString("en-GB", { month: "long", year: "numeric" });
}

File: backend/app/core/currency.py
"""Currency formatting (§24).

Transactions are always stored and displayed in their ORIGINAL currency; this
module only formats, it never converts. Indian-style lakh/crore grouping is
applied for INR because that is how prices are actually quoted there.
"""
from __future__ import annotations

SYMBOLS: dict[str, str] = {
    "GBP": "£", "EUR": "€", "USD": "$", "AUD": "A$", "CAD": "C$",
    "NZD": "NZ$", "INR": "₹", "JPY": "¥", "CHF": "CHF ", "SEK": "kr ",
    "NOK": "kr ", "DKK": "kr ", "ZAR": "R", "BRL": "R$", "SGD": "S$",
    "HKD": "HK$", "PLN": "zł ", "CZK": "Kč ", "HUF": "Ft ", "RON": "lei ",
    "ISK": "kr ", "TRY": "₺", "RSD": "din ", "UAH": "₴", "KRW": "₩",
    "CNY": "¥", "MXN": "MX$", "AED": "AED ", "ILS": "₪",
}

# ISO 3166-1 alpha-2 -> ISO 4217. Used when a provider does not state one.
#
# Completeness matters here: an incomplete map used to fall through to a
# hard-coded "EUR", which labelled Hungarian figures in euros. A missing entry
# now yields None and the currency is simply not shown, rather than asserting a
# currency the country does not use.
COUNTRY_CURRENCY: dict[str, str] = {
    # --- euro area ---------------------------------------------------------
    "AT": "EUR", "BE": "EUR", "CY": "EUR", "DE": "EUR", "EE": "EUR",
    "ES": "EUR", "FI": "EUR", "FR": "EUR", "GR": "EUR", "HR": "EUR",
    "IE": "EUR", "IT": "EUR", "LT": "EUR", "LU": "EUR", "LV": "EUR",
    "MT": "EUR", "NL": "EUR", "PT": "EUR", "SI": "EUR", "SK": "EUR",
    # Bulgaria adopted the euro on 1 January 2026. Index series published
    # before then were denominated in leva; since an index has no price level
    # this affects labelling only.
    "BG": "EUR",
    # --- rest of Europe ----------------------------------------------------
    "GB": "GBP", "CH": "CHF", "CZ": "CZK", "DK": "DKK", "HU": "HUF",
    "IS": "ISK", "NO": "NOK", "PL": "PLN", "RO": "RON", "SE": "SEK",
    "TR": "TRY", "RS": "RSD", "UA": "UAH",
    # --- rest of world -----------------------------------------------------
    "US": "USD", "CA": "CAD", "AU": "AUD", "NZ": "NZD", "IN": "INR",
    "JP": "JPY", "ZA": "ZAR", "BR": "BRL", "SG": "SGD", "HK": "HKD",
    "KR": "KRW", "CN": "CNY", "MX": "MXN", "AE": "AED", "IL": "ILS",
}


def symbol(currency: str) -> str:
    return SYMBOLS.get(currency.upper(), currency.upper() + " ")


def currency_for_country(iso2: str | None) -> str | None:
    if not iso2:
        return None
    return COUNTRY_CURRENCY.get(iso2.upper())


def format_full(value: float, currency: str) -> str:
    """`£425,000` — the exact figure, for headline values."""
    return f"{symbol(currency)}{round(value):,}"


def format_compact(value: float, currency: str) -> str:
    """`£725k`, `£1.2m`, `₹1.8 Cr` — for map markers (§8, §24)."""
    cur = currency.upper()
    sym = symbol(cur)
    v = float(value)

    if cur == "INR":
        # Indian numbering: 1 crore = 10,000,000; 1 lakh = 100,000.
        if v >= 10_000_000:
            return f"{sym}{_trim(v / 10_000_000)} Cr"
        if v >= 100_000:
            return f"{sym}{_trim(v / 100_000)} L"
        return f"{sym}{round(v):,}"

    if cur == "JPY":
        if v >= 100_000_000:
            return f"{sym}{_trim(v / 100_000_000)}億"
        if v >= 10_000:
            return f"{sym}{_trim(v / 10_000)}万"
        return f"{sym}{round(v):,}"

    if v >= 1_000_000_000:
        return f"{sym}{_trim(v / 1_000_000_000)}bn"
    if v >= 1_000_000:
        return f"{sym}{_trim(v / 1_000_000)}m"
    if v >= 1_000:
        return f"{sym}{round(v / 1_000)}k"
    return f"{sym}{round(v):,}"


def _trim(x: float) -> str:
    """One decimal place, but drop a trailing `.0`."""
    s = f"{x:.1f}"
    return s[:-2] if s.endswith(".0") else s
